### PR TITLE
AFS and branch length cleanups

### PR DESCRIPTION
**Pull request by jeromekelleher**
This  should clears up any remaining issues with the AFS. It does feel weird not to do the same thing for the site AFS, but perhaps the same issues don't occur?  Any thoughts @petrelaharp?

### CODE REVIEW

**Comment by petrelharp** [comment]
Thanks; I think there's two things still though:
1. We should do the same thing for site and branch.
2. We're still folding wrongly (discussed over [here](https://github.com/tskit-dev/tskit/issues/203#issuecomment-513871058) - there's a proposed solution there.)

**Comment by petrelharp** [comment]
To be clear, the proposed solution is, when folding, to output an array with the *same* dimensions as unfolded, but with half the entries zero. My proposal is to fold so that the counts given are of the *minor* allele; so the algorithm would check if `k1+k2+...+km > (n1+n2+...+nm)/2`, and if so, set `AFS[n1-k1,n2-k2,...,nm-km] += AFS[k1,k2,...,km]` and then `AFS[n1-k1,n2-k2,...,nm-km] = 0`. Sorry for not figuring this out earlier.

**Comment by jeromekelleher** [comment]
Sounds good @petrelharp. Keeping the dimensions the same is a major win in my book. I'll code this up and get back to you.

**Comment by jeromekelleher** [comment]
I've updated the semantics of the site AFS here @petrelharp, would you mind taking a look please? The basic idea is not to include anything that's monomorphic across all samples. I've not pushed the changes through the C yet.

This brings up (yet another) subtle issue. In the branch AFS, we're excluding all branches that are ancestral to 0 samples, but including all branches ancestral to n samples (which is inconsistent with the site definition and to the ``segregating_sites`` function. Should we change this to also remove these branches that are ancestral to all samples?

I'm close to AFS'd out here - these corner cases are never ending!

**Comment by jeromekelleher** [comment]
@petrelharp --- see the last commit: is this what you have in mind for folding? (I'll push through the rest of the changes, I just wanted to make sure I'm not on the wrong path first).)

**Comment by petrelharp** [comment]
Er, no, that's not quite right. And that was extremely annoying to figure out, but I got it.
```
import numpy as np
import itertools

def fold(x, dims):
    """
    Folds the specified coordinates.
    """
    x = np.array(x)
    dims = np.array(dims)
    k = len(dims)
    n = np.sum(dims - 1) / 2
    s = np.sum(x)
    while s == n and k >= 0:
        k -= 1
        n -= (dims[k] - 1)/2
        s -= x[k]
    if s > n:
        x = dims - 1 - x
    assert np.all(x >= 0)
    return tuple(x)


def foldit(A):
    O = np.zeros(A.shape)
    dims = A.shape
    inds = [range(k) for k in dims]
    for ij in itertools.product(*inds):
        nij = fold(ij, dims)
        O[nij] += A[ij]
    return O


A = np.arange(12)
Af = np.array([11., 11., 11., 11., 11., 11.,  0.,  0.,  0.,  0.,  0.,  0.])
B = A.copy().reshape(3,4)
Bf = np.array([[11., 11., 11.,  0.],
               [11., 11.,  0.,  0.],
               [11.,  0.,  0.,  0.]])
C = A.copy().reshape(3,2,2)
Cf = np.array([[[11., 11.],
                [11., 11.]],
               [[11., 11.],
                [ 0.,  0.]],
               [[ 0.,  0.],
                [ 0.,  0.]]])
D = np.arange(9).reshape((3,3))
Df = np.array([[8., 8., 8.],
               [8., 4., 0.],
               [0., 0., 0.]])
E = np.arange(9)
Ef = np.array([8., 8., 8., 8., 4., 0., 0., 0., 0.])

assert np.all(foldit(A) == Af)
assert np.all(foldit(B) == Bf)
assert np.all(foldit(C) == Cf)
assert np.all(foldit(D) == Df)
assert np.all(foldit(E) == Ef)
```

**Comment by jeromekelleher** [comment]
Note to self: need to do this after folding is sorted:

> This brings up (yet another) subtle issue. In the branch AFS, we're excluding all branches that are ancestral to 0 samples, but including all branches ancestral to n samples (which is inconsistent with the site definition and to the segregating_sites function. Should we change this to also remove these branches that are ancestral to all samples?

**Comment by petrelharp** [comment]
>   This brings up (yet another) subtle issue. In the branch AFS, we're excluding all branches that are ancestral to 0 samples, but including all branches ancestral to n samples (which is inconsistent with the site definition and to the segregating_sites function. Should we change this to also remove these branches that are ancestral to all samples?

We should definately do the same thing for both branch and site. My inclination would be to *include* the branches/mutations ancestral to all samples in both (otherwise the last entry in the AFS is always 0?). Number of fixations seems like a nice thing to compute if they're actually in the tree sequence.

**Comment by petrelharp** [comment]
> I'm close to AFS'd out here

seriously.

**Comment by jeromekelleher** [comment]
> seriously.

Yah, what a can of worms...

**Comment by jeromekelleher** [comment]
OK, that's it, I am done with the AFS (tempted rename it the FFS). As far as I know, site and branch are consistent now and should deal gracefully with arbitrarily weird stuff.

> We should definately do the same thing for both branch and site. My inclination would be to include the branches/mutations ancestral to all samples in both (otherwise the last entry in the AFS is always 0?). Number of fixations seems like a nice thing to compute if they're actually in the tree sequence.

I made this change @petrelharp, and some annoying corner cases bubbled up, which significantly reduced my will to live. So, I reverted and pushed the change through where we only update if 0 < all_samples_count < n in both site and branch AFS. This is, at least, consistent with the definition of segregating sites. 

Feel free to change it if you like though, I really don't have a strong opinion. It needs documenting now, but this will have to wait for another day I think.

**Comment by jeromekelleher** [comment]
So yeah, I guess we merge this much and maybe get Konrad to try it out and see if he likes your folding mechanism? I can push out another alpha to make the code available.

**Comment by petrelharp** [comment]
This is great. I agree, let's merge it. I've udpated the documentation. I would like in principle to get around to putting the fixed sites in; we'll see.

Hm, one thing: I thought we agreed to set the default to `polarised=True` for the AFS?

**Comment by petrelharp** [comment]
Hm, that Travis error might be because I erroneously pushed to a new branch called "pull/274/head", and then deleted it, on the repository. :oh-no: :whoops-face:

File: python/tskit/trees.py [hunk 1]
```diff
@@ -3812,6 +3813,10 @@ def allele_frequency_spectrum(
         "node"
             Not supported for this method (raises a ValueError).
 
+        The method does *not* currently count the number of fixed alleles
```
Comment (jeromekelleher) [inline python/tskit/trees.py]: This is overstating it:
```python
ts = msprime.simulate(6, random_seed=2)
print(ts.draw_text())
afs = ts.allele_frequency_spectrum([[0, 2]], mode="branch", polarised=True)                                                  
print(afs)
```
gives

```
              10      ┊  
    ┊   ┏━━┻━━┓   ┊  
1.06┊   ┃     9   ┊  
    ┊   ┃   ┏━┻┓  ┊  
0.80┊   8   ┃  ┃  ┊  
    ┊ ┏━┻┓  ┃  ┃  ┊  
0.61┊ ┃  7  ┃  ┃  ┊  
    ┊ ┃ ┏┻┓ ┃  ┃  ┊  
0.08┊ ┃ ┃ ┃ ┃  6  ┊  
    ┊ ┃ ┃ ┃ ┃ ┏┻┓ ┊  
0.00┊ 3 0 2 1 4 5 ┊  
  0.00          1.00 

[[3.45905848 1.22532517 0.90607714]]

```
It's only stuff above the MRCA of **all** samples that's ignored, so the final entry will have values when the sample sets are a subset of all samples.

**Comment by jeromekelleher** [comment]
> This is great. I agree, let's merge it. I've udpated the documentation. I would like in principle to get around to putting the fixed sites in; we'll see.

Sure, I'm happy to for this to be changed. See my comment about though: the final value of the AFS is used currently. 

> Hm, one thing: I thought we agreed to set the default to polarised=True for the AFS?

So, there's a technical issue with doing this at the C level: TSK_STAT_POLARISED is a bit flag and we can't default it to being set. Also, it seems a bit confusing to have polarised the default in some places but not in others. So, my take at this point would be to either leave things as they are, or to change the signature of this method to replace ``polarised=False`` with ``folded=False``, and document that folded == (not polarised).

**Comment by petrelharp** [comment]
> So, my take at this point would be to either leave things as they are, or to change the signature of this method to replace polarised=False with folded=False, and document that folded == (not polarised).

Ok, let's leave things as they are.

File: python/tskit/trees.py (hunk 1 above)
Comment (petrelharp) [reply python/tskit/trees.py]: Good point; I've fixed this.

**Comment by jeromekelleher** [comment]
> Ok, let's leave things as they are

Done! Merging; hooray no more FFS!

**Comment by jeromekelleher** [comment]
Thanks for your help BTW @petrelharp  - I think I'd have just dropped this a while ago otherwise!

**Comment by petrelharp** [comment]
Thank YOU for doing the work, and sorry again for misleading on the folding...

**Pull request merged**